ouroboros-consensus/docs/GenesisDecomposition.md

```diff
@@ -127,9 +136,190 @@ in 2021 March using chapter 21 of the Consensus Report.
+        fetch. EG 2 peers bulk fetchs quite a bit faster than 10 peers.
+
+      * The suggested delta is the exact opposite of the plan above: when
+        catching up, we need many many peers in order to increase the chance
+        that at least one has the honest chain. (We haven't considered actual
+        counts rigourously.)
+
+      * Mitigation plan: the ChainSync clients need to share some work, so that
+        we don't revalidate the same header once-per-peer.
+
+      * This violates avg-case = worst-case, but that's OK when bulk syncing.
+        Perhaps we should guard this work sharing by `NotIsAlert`.
+
   * SEPARATE WORK:
 
       * TODO potential optimizations (Section 21.6.3)
 
       * TODO testing etc - we'd _very much really_ like to use the ThreadNet
         rewrite for this
+
+-----
+
+Updated on 2021 August 9, after much additional thought and broader
+reconsiderations, kicked off by Javier Sagredo's observation of a stalling
+attack vector in the original sketch above.
+
+This new sketch updates much-but-not-all of the origial sketch above.
+
+- Execution begins in the _Syncing_ state.
+
+- While we are Syncing:
+
+    - If our valency falls below some threshold, then BlockFetch stops sending
+      new fetch requests until sufficient valency is recovered.
+
+    - BlockFetch can only download blocks from the headers that the density
+      rule approves.
+
+        - The density rule is: compare header chains based on the number of
+          headers in the relevant Genesis window (the 3k/f slots after the
+          intersection), though if the headers do not span the Genesis window
+          and the peer claims to have more headers we must wait for them
+          (because they might also be in the window).
+
+        - The Ouroboros Genesis paper proves -- excepting only disasterous
+          intervals -- that density rule will always strictly prefer the honest
+          chain over any possible alternative.
+
+    - Therefore, we require that each peer's highwater blockno is increasing
+      "fast enough on average" until we're at their tip, with the only
+      exceptional circumstance being when their latest header is beyond our
+      forecast range (since we don't even request a next header while that is
+      true).
+
+        - TODO Do we actually need that exception? Under what circumstances
+          would it be relevant, during Syncing?
+
+        - TODO I'm anticipating a token bucket for enforcing "fast enough on
+          average", but there remain plenty of details and thresholds to
+          consider.
+
+        - A possible refinement: if they can promise to send a specific k+1st
+          block (which the honest nodes would always do, up to their immutable
+          tip), then they're allowed to be somewhat slower, since we'll
+          disconnect from them if either they don't deliver that block or if
+          the eventual densest chain does not include that block.
```

Comment:
I assume this is the case but I want to confirm it. We are going to disconnect from adversarial peers and then wait until we reconnect to enough peers to continue?